materialize-databricks/sqlgen.go

```diff
@@ -0,0 +1,269 @@
+	{{- if $ind }},{{ end }}
+	{{ $key.Identifier }} {{ $key.DDL }}
+{{- end }}
+);
+{{ end }}
+
+-- Templated truncation of the temporary load table:
+{{ define "truncateLoadTable" }}
+TRUNCATE TABLE {{ template "temp_name_load" . }};
+{{ end }}
+
+{{ define "dropLoadTable" }}
+DROP TABLE IF EXISTS {{ template "temp_name_load" . }}
+{{ end }}
+
+-- Idempotent creation of the store table for staging new records.
+{{ define "createStoreTable" }}
+CREATE TABLE IF NOT EXISTS {{ template "temp_name_store" . }};
+{{ end }}
+
+-- Templated truncation of the temporary store table:
+{{ define "truncateStoreTable" }}
+TRUNCATE TABLE {{ template "temp_name_store" . }};
+{{ end }}
+
+{{ define "dropStoreTable" }}
+DROP TABLE IF EXISTS {{ template "temp_name_store" . }}
+{{ end }}
+
+-- Templated creation of a materialized table definition and comments:
+{{ define "createTargetTable" }}
+CREATE TABLE IF NOT EXISTS {{$.Identifier}} (
+  {{- range $ind, $col := $.Columns }}
+  {{- if $ind }},{{ end }}
+  {{$col.Identifier}} {{$col.DDL}} COMMENT {{ Literal $col.Comment }}
+  {{- end }}
+) COMMENT {{ Literal $.Comment }};
+{{ end }}
+
+-- Templated query which joins keys from the load table with the target table, and returns values. It
+-- deliberately skips the trailing semi-colon as these queries are composed with a UNION ALL.
+
+{{ define "loadQuery" }}
+{{ if $.Table.Document -}}
+SELECT {{ $.Table.Binding }}, {{ $.Table.Identifier }}.{{ $.Table.Document.Identifier }}
+	FROM {{ $.Table.Identifier }}
+	JOIN {{ template "temp_name_load" . }} AS r
+	ON {{ range $ind, $key := $.Table.Keys }}
+	{{- if $ind }} AND {{ end -}}
+	{{ $.Table.Identifier }}.{{ $key.Identifier }} = r.{{ $key.Identifier }}
+	{{- end }}
+{{ else -}}
+SELECT -1, ""
+{{ end -}}
+{{ end }}
+
+{{ define "cast" }}
+{{- if Contains $ "DATE" -}}
+	::DATE
+{{- else if Contains $ "TIMESTAMP" -}}
+	::TIMESTAMP
+{{- else if Contains $ "DOUBLE" -}}
+	::DOUBLE
+{{- else if Contains $ "BIGINT" -}}
+	::BIGINT
+{{- else if Contains $ "BOOLEAN" -}}
+	::BOOLEAN
+{{- else if Contains $ "BINARY" -}}
+	::BINARY
+{{- end }}
+{{ end }}
+
+-- Directly copy into the target table
+{{ define "copyIntoDirect" }}
+	COPY INTO {{ $.Table.Identifier }} FROM (
+    SELECT
+		{{ range $ind, $key := $.Table.Columns }}
+			{{- if $ind }}, {{ end -}}
+			{{$key.Identifier -}}{{ template "cast" $key.DDL }}
+		{{- end }}
+  FROM {{ Literal $.StagingPath }}
+	)
+  FILEFORMAT = JSON
+  FILES = (%s)
+  FORMAT_OPTIONS ( 'mode' = 'FAILFAST' )
+  ;
+{{ end }}
+
+-- Copy into temporary store table
+{{ define "copyIntoStore" }}
+	COPY INTO {{ template "temp_name_store" . }} FROM (
+    SELECT
+		_metadata.file_name as _metadata_file_name,
+		{{ range $ind, $key := $.Table.Columns }}
+			{{- if $ind }}, {{ end -}}
+			{{$key.Identifier -}}
+		{{- end }}
+    FROM {{ Literal $.StagingPath }}
+	)
+  FILEFORMAT = JSON
+  FILES = (%s)
+  FORMAT_OPTIONS ( 'mode' = 'FAILFAST', 'mergeSchema' = 'true', 'primitivesAsString' = 'true' )
```

Comment:
@williamhbaker since we will do a casting to the target table's type during copying, we do not need databricks to infer numerical or boolean values for us (that's what `primitive` refers to mostly), we can keep them as string. The issue is that if we do ask databricks to infer primitive types for us, it will not always be right, for example a column might be a float, but if databricks sees only whole numbers in one transaction, it will infer the type as int, so in the end we do need to cast to make sure we get consistent types and avoid data type mismatch errors.

`mergeSchema` allows the schema of source files to influence each other: in cases where we are loading from multiple files, this option tells databricks to try to come up with a schema that matches _all_ of the files, rather than using independent inferred schemas for each file

Reply:
Is there a way to completely specify the schema with the query and/or disabled inference entirely? Our MotherDuck materialization does that [example](https://github.com/estuary/connectors/blob/57d097e5f49b570ed12d2b07d9ad7b64bd94f319/materialize-motherduck/.snapshots/TestSQLGeneration#L14-L17), and I found it sped things up a lot. Maybe that's what the casting is doing?

Reply:
@williamhbaker I have tried multiple ways to ask databricks to do this, but have not been successful. I just pushed some changes which make the schema of the temporary tables explicit as well, but still, during the copy, I get errors like this:

```
transactor.Store: store: copying into to temporary table: databricks: execution error: failed to execute query: unexpected operation state ERROR_STATE: A schema mismatch detected when writing to the Delta table (Tabl
e ID: 5a80f00b-1321-4686-a285-b83e37a2c905).                                                                                                                                                                            
To enable schema migration using DataFrameWriter or DataStreamWriter, please set:                                                                                                                                       
'.option("mergeSchema", "true")'.                                                                                                                                                                                       
For other operations, set the session configuration                                                                                                                                                                     
spark.databricks.delta.schema.autoMerge.enabled to "true". See the documentation                                                                                                                                        
specific to the operation for details.                                                                                                                                                                                  
                                                                                                                                                                                                                        
Table schema:                                                                                                                                                                                                           
root                                                                                                                                                                                                                    
-- id: long (nullable = false)                                                                                                                                                                                          
-- array_int: string (nullable = true)                                                                                                                                                                                  
-- bool_field: boolean (nullable = true)                                                                                                                                                                                
-- float_field: double (nullable = true)                                                                                                                                                                                
-- flow_published_at: timestamp (nullable = false)                                                                                                                                                                      
-- multiple: string (nullable = true)                                                                                                                                                                                   
-- nested: string (nullable = true)                                                                                                                                                                                     
-- nullable_int: long (nullable = true)                                                                                                                                                                                 
-- str_field: string (nullable = false)                                                                                                                                                                                 
-- flow_document: string (nullable = false)                                                                                                                                                                             
                                                                                                                                                                                                                        
                                                                                                                                                                                                                        
Data schema:                                                                                                                                                                                                            
root                                                                                                                                                                                                                    
-- _metadata_file_name: string (nullable = true)                                                                                                                                                                        
-- id: long (nullable = true)                                                                                                                                                                                           
-- array_int: string (nullable = true)                                                                                                                                                                                  
-- bool_field: boolean (nullable = true)                                                                                                                                                                                
-- float_field: double (nullable = true)                                                                                                                                                                                
-- flow_published_at: timestamp (nullable = true)                                                                                                                                                                       
-- multiple: string (nullable = true)                                                                                                                                                                                   
-- nested: string (nullable = true)                                                                                                                                                                                     
-- nullable_int: long (nullable = true)                                                                                                                                                                                 
-- str_field: string (nullable = true)                                                                                                                                                                                  
-- flow_document: string (nullable = true) 
```

In this instance it seems even whether a column is nullable or not is causing issues... I would have loved it if Databricks could have been smarter about this.

Reply:
Ah, oh well, thanks for checking & for the explanation!